src/ruis/render/scene/gltf_loader.cpp

```diff
@@ -0,0 +1,772 @@
+		}
+	}
+
+	// std::cout << "loading accessors" << std::endl;
+	it = json.object().find("accessors");
+	if (it != json.object().end() && it->second.is_array()) {
+		for (const auto& sub_json : it->second.array()) {
+			accessors.push_back(read_accessor(sub_json));
+		}
+	}
+
+	// std::cout << "loading meshes" << std::endl;
+	it = json.object().find("meshes");
+	if (it != json.object().end() && it->second.is_array()) {
+		for (const auto& sub_json : it->second.array()) {
+			meshes.push_back(read_mesh(sub_json));
+		}
+	}
+
+	// std::cout << "loading nodes" << std::endl;
+	it = json.object().find("nodes");
+	if (it != json.object().end() && it->second.is_array()) {
+		for (const auto& sub_json : it->second.array()) {
+			nodes.push_back(read_node(sub_json));
+		}
+	}
+
+	// hierarchize nodes
+	ASSERT(nodes.size() == child_indices.size())
+	for (uint32_t i = 0; i < nodes.size(); ++i) {
+		for (uint32_t ci : child_indices[i]) {
+			nodes[i].get().children.push_back(nodes[ci]);
+		}
+	}
+
+	// std::cout << "loading scenes" << std::endl;
+	it = json.object().find("scenes");
+	if (it != json.object().end() && it->second.is_array()) {
+		for (const auto& sub_json : it->second.array()) {
+			scenes.push_back(read_scene(sub_json));
+		}
+	}
+
+	// create scene
+	auto active_scene = utki::make_shared<scene>();
+	int active_scene_index = read_int(json, "scene");
+	if (active_scene_index < 0) {
+		// this .gltf file is a library
+	} else {
+		active_scene = scenes[active_scene_index];
+	}
+
+	// create cameras (currently generates one default camera)
+	// //std::cout << "create camera" << std::endl;
+
+	// auto cam = utki::make_shared<camera>();
+	// cam.get().pos = {0, 3.5, -8};
+	// cam.get().target = {0, 1, 0};
+	// cam.get().up = {0, 1, 0};
+	// cam.get().near = 0.1;
+	// cam.get().far = 100;
+	// cam.get().fovy = 3.1415926 / 3;
+
+	// active_scene.get().cameras.push_back(cam);
+	// active_scene.get().active_camera = cam.to_shared_ptr();
+
+	// std::cout << "gltf load finished" << std::endl;
+	return active_scene;
+}
+
+template <typename tp_type>
+utki::shared_ref<ruis::render::vertex_array> gltf_loader::create_vao_with_tangent_space(
+	utki::shared_ref<accessor> index_accessor,
+	utki::shared_ref<accessor> position_accessor,
+	utki::shared_ref<accessor> texcoord_0_accessor,
+	utki::shared_ref<accessor> normal_accessor
+)
+{
+	uint32_t total_vertices = position_accessor.get().count;
+	uint32_t total_triangles = index_accessor.get().count / 3;
+
+	// std::cout << "index is " << sizeof(tp_type) << "-byte" << std::endl;
+
+	const auto& indices = std::get<std::vector<tp_type>>(index_accessor.get().data);
+	const auto& positions = std::get<std::vector<ruis::vec3>>(position_accessor.get().data);
+	const auto& texcoords = std::get<std::vector<ruis::vec2>>(texcoord_0_accessor.get().data);
+	const auto& normals = std::get<std::vector<ruis::vec3>>(normal_accessor.get().data);
+
+	std::vector<ruis::vec3> tangents; // or vec4 ?
```

Comment:
is still under question vec3 or vec4? Remove comment if vec3 for sure.

Reply:
removed